quantmsio/core/psm.py

```
@@ -1,329 +1,130 @@
-        global_qvalue = (
-            None if ("global_qvalue" not in psm or psm["global_qvalue"] is None) else np.float64(psm["global_qvalue"])
+    def slice(df, partitions):
+        cols = df.columns
+        if not isinstance(partitions, list):
+            raise Exception(f"{partitions} is not a list")
+        if len(partitions) == 0:
+            raise Exception(f"{partitions} is empty")
+        for partion in partitions:
+            if partion not in cols:
+                raise Exception(f"{partion} does not exist")
+        for key, df in df.groupby(partitions):
```

Ruff

Loop control variable `df` overrides iterable it iterates

(B020)